File: src/main/java/com/fatec/destino/repository/usuario/CompraRepository.java
package com.fatec.destino.repository.usuario;

import com.fatec.destino.model.Compra;
import com.fatec.destino.model.pacote.Pacote;
import com.fatec.destino.model.usuario.Usuario;
import com.fatec.destino.util.model.pacote.PacoteStatus;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.data.jpa.repository.Query;
import org.springframework.data.repository.query.Param;
import org.springframework.stereotype.Repository;

import java.util.List;
import java.util.Optional;
import java.util.UUID;

@Repository
public interface CompraRepository extends JpaRepository<Compra, Long> {

    // Agrupa compras por Mês. Retorna [Mês (Integer), Quantidade (Long)]
    @Query("SELECT MONTH(c.dataCompra), COUNT(c) FROM Compra c WHERE YEAR(c.dataCompra) = :ano GROUP BY MONTH(c.dataCompra) ORDER BY MONTH(c.dataCompra) ASC")
    List<Object[]> countComprasByMonth(@Param("ano") int ano);

    @Query("SELECT c FROM Compra c WHERE c.usuario.id = :usuarioId ORDER BY c.pacote.inicio DESC")
    List<Compra> findAllByUsuarioId(@Param("usuarioId") UUID usuarioId);

    @Query("SELECT c FROM Compra c WHERE c.usuario.id = :usuarioId AND c.pacote.status = :status ORDER BY c.pacote.inicio DESC")
    List<Compra> findAllByUsuarioIdWhereStatusConcluido(@Param("usuarioId") UUID usuarioId, @Param("status") PacoteStatus status);

    @Query("SELECT c FROM Compra c WHERE c.usuario.id = :usuarioId AND c.pacote.status = :status ORDER BY c.pacote.inicio DESC")
    List<Compra> findAllByUsuarioIdWhereStatusEmAdamento(@Param("usuarioId") UUID usuarioId, @Param("status") PacoteStatus status);

    // Busca compra específica garantindo que pertence ao usuario (segurança)
    @Query("SELECT c FROM Compra c WHERE c.id = :compraId AND c.usuario.email = :emailUsuario")
    Compra findByIdAndUsuarioEmail(@Param("compraId") Long compraId, @Param("emailUsuario") String emailUsuario);

    // Em CompraRepository.java
    Optional<Compra> findByUsuarioAndPacote(Usuario usuario, Pacote pacote);
}

File: src/main/kotlin/com/fatec/destino/repository/usuario/CompraRepository.kt
package com.fatec.destino.repository.usuario

import com.fatec.destino.model.Compra
import com.fatec.destino.model.pacote.Pacote
import com.fatec.destino.model.usuario.Usuario
import com.fatec.destino.util.model.pacote.PacoteStatus
import com.fatec.destino.util.repository.CompraMesProjection
import org.springframework.data.jpa.repository.JpaRepository
import org.springframework.data.jpa.repository.Query
import org.springframework.data.repository.query.Param
import org.springframework.stereotype.Repository
import java.util.*

@Repository
interface CompraRepository : JpaRepository<Compra, Long> {
    @Query("""
        SELECT MONTH(c.dataCompra) as mes, COUNT(c) as total 
        FROM Compra c 
        WHERE YEAR(c.dataCompra) = :ano 
        GROUP BY MONTH(c.dataCompra) 
        ORDER BY MONTH(c.dataCompra) ASC
    """)
    fun countComprasByMonth(@Param("ano") ano: Int): List<CompraMesProjection>

    @Query("SELECT c FROM Compra c WHERE c.usuario.id = :usuarioId ORDER BY c.pacote.inicio DESC")
    fun findAllByUsuarioId(@Param("usuarioId") usuarioId: UUID?): MutableList<Compra?>?

    @Query("SELECT c FROM Compra c WHERE c.usuario.id = :usuarioId AND c.pacote.status = :status ORDER BY c.pacote.inicio DESC")
    fun findAllByUsuarioIdWhereStatusConcluido(
        @Param("usuarioId") usuarioId: UUID?,
        @Param("status") status: PacoteStatus?
    ): MutableList<Compra?>?

    @Query("SELECT c FROM Compra c WHERE c.usuario.id = :usuarioId AND c.pacote.status = :status ORDER BY c.pacote.inicio DESC")
    fun findAllByUsuarioIdWhereStatusEmAdamento(
        @Param("usuarioId") usuarioId: UUID?,
        @Param("status") status: PacoteStatus?
    ): MutableList<Compra?>?

    // Busca compra específica garantindo que pertence ao usuario (segurança)
    @Query("SELECT c FROM Compra c WHERE c.id = :compraId AND c.usuario.email = :emailUsuario")
    fun findByIdAndUsuarioEmail(
        @Param("compraId") compraId: Long?,
        @Param("emailUsuario") emailUsuario: String?
    ): Compra?

    // Em CompraRepository.java
    fun findByUsuarioAndPacote(usuario: Usuario?, pacote: Pacote?): Optional<Compra?>?
}